torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,159 @@ namespace rpc {
 
 constexpr size_t WorkerInfo::MAX_NAME_LEN;
 
+// kRpcRetryDuration is the initial duration between consecutive RPC retries.
+constexpr auto kRpcRetryDuration = std::chrono::milliseconds(1000);
+// kBackoff is the multiplier for the exponential backoff algorithm.
+constexpr float kBackoff = 1.5;
+
 RpcAgent::RpcAgent(
     WorkerInfo workerId,
     std::unique_ptr<RequestCallback> cb,
     std::chrono::milliseconds rpcTimeout)
     : workerInfo_(std::move(workerId)),
       cb_(std::move(cb)),
       rpcTimeout_(rpcTimeout),
-      profilingEnabled_(false) {}
+      profilingEnabled_(false) {
+  rpcAgentRunning_.store(true);
+  rpcRetryThread_ = std::thread(&RpcAgent::retryExpiredRpcs, this);
+}
+
+RpcAgent::~RpcAgent() {
+  if (rpcAgentRunning_.load()) {
+    shutdown();
+  }
+}
+
+void RpcAgent::shutdown() {
+  if (!rpcAgentRunning_.exchange(false)) {
+    return;
+  }
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
+  shutdownImpl();
+}
+
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  TORCH_CHECK(maxRetries >= 0, "maxRetries cannot be negative.");
+  auto originalFuture = std::make_shared<FutureMessage>();
+  steady_clock_time_point newTime = computeNewRpcRetryTime(/* retryCount */ 0);
+  Message& msg = message;
+  auto fm = send(to, std::move(message));
+  auto firstRetryRpc = std::make_shared<RpcRetryInfo>(
+      to, std::move(msg), originalFuture, /* retryCount */ 0, maxRetries);
+
+  fm->addCallback([this, newTime, firstRetryRpc](
+                      const rpc::Message& lambdaMessage,
+                      const c10::optional<utils::FutureError>& futErr) {
+    rpcRetryCallback(lambdaMessage, futErr, newTime, firstRetryRpc);
+  });
+
+  return originalFuture;
+}
+
+void RpcAgent::retryExpiredRpcs() {
+  while (rpcAgentRunning_.load()) {
+    std::unique_lock<std::mutex> lock(rpcRetryMutex_);
+
+    // We must continue sleeping as long as the RPC Agent is running and when
+    // either the Retry Map is empty, or when the Retry Map's earliest expiring
+    // RPC is set to be retried in the future.
+    if (rpcRetryMap_.empty()) {
+      rpcRetryMapCV_.wait(lock, [&, this] {
+        return !(rpcAgentRunning_.load() && rpcRetryMap_.empty());
+      });
+    }
+
+    steady_clock_time_point earliestTimeout;
 
-RpcAgent::~RpcAgent() = default;
+    if (!rpcRetryMap_.empty()) {
+      earliestTimeout = rpcRetryMap_.begin()->first;
+      rpcRetryMapCV_.wait_until(lock, earliestTimeout, [&, this] {
```

Comment:
It seems like we should be able to combine this wait_until with the previous wait, with sufficient lambda condition.

Reply:
@jjlilley  could you please elaborate on how these two can be combined? Looks like they can be put into a if-else clause, so that we only need to wait once, but not sure if we can use a single wait/wait_until to achieve that.

Reply:
@mrshenli Sure- it's probably easier to express with the traditional, non-lambda calling convention-

I think something like the following would unify the condvar waiting (maybe I'm missing something though?)

   for (;;) {
      if (!rpcAgentRunning_.load()) break;
       auto now = std::chrono::steady_clock::now();
       auto earliestTimeout = rpcRetryMap_.empty() ? now + kManyYears : rpcRetryMap_.begin()->first;
       if (now >= earliestTimeout) break;
       rpcRetryMapCV_.wait_until(lock, earliestTimeout);
    }

IMO, unifying the waits tends to make the code easier to maintain over time, though there could be other considerations here.

Reply:
@jjlilley I tried that snippet, and played around with a few more potential solutions for unifying the condvar waiting as much as possible. The above snippet works if we set an arbitrarily large duration for `kManyYears`, for example `std::chrono::hours(10000)`, but the tests time out when using `std::chrono::time_point::max()`. I can drill down on this further and create an issue to track, if using a large constant for the wait duration for the empty map is not preferred. An alternative is as follows:

```
auto earliestTimeout = std::chrono::time_point<std::chrono::steady_clock>::max();
for (;;) {
  if (!rpcAgentRunning_.load()) break;
  if (std::chrono::steady_clock::now() >= earliestTimeout) break;
  if (rpcRetryMap_.empty()) {
    rpcRetryMapCV_.wait(lock);
  } else {
    earliestTimeout = rpcRetryMap_.begin()->first;
    rpcRetryMapCV_.wait_until(lock, earliestTimeout);
  }
}
```
This doesn't consolidate into a single condvar wait, however. I'm happy to go with whichever one you feel is better coding standards-wise!

Reply:
According to (https://stackoverflow.com/questions/27726818/stdcondition-variablewait-for-exits-immediately-when-given-stdchronodura), using the max time_point with condvar wait_until is a known issue. Seems like the implementation in `ProcessGroupAgent::pollTimedOutRPCs()` on line 528 does `wait_until` if the expiration time is not max, otherwise it uses `wait`.